File: apps/api/app/schemas/credit_file.py
from datetime import datetime
from enum import StrEnum
from typing import Literal

from pydantic import BaseModel

from app.agents.schemas import TraceStep
from app.schemas.audit import AuditEventSchema
from app.schemas.msme import MSMEDetail
from app.schemas.portfolio import PortfolioCase
from app.schemas.prospect import ProspectSignalOutputSchema
from app.schemas.score import EarlyWarningTrigger, ScoreOutputSchema
from app.services.transaction_summary_service import TransactionSummary


class EvidenceStatusItem(BaseModel):
    source_type: str
    source_label: str
    status: str
    why_it_matters: str
    related_score_component: str
    action_label: str
    action_enabled: bool
    disabled_reason: str | None = None


class EvidenceExtractedSignal(BaseModel):
    field_name: str
    value: str
    source_mapping: str
    confidence: int


class EvidenceType(StrEnum):
    bank_statement = "bank_statement"
    gst_returns = "gst_returns"
    bureau_report = "bureau_report"
    itr = "itr"
    udyam = "udyam"
    uploaded_document = "uploaded_document"
    gem_profile = "gem_profile"
    identity_proof = "identity_proof"
    address_proof = "address_proof"
    other = "other"


class EvidenceRecord(BaseModel):
    id: str
    msme_id: str
    evidence_type: EvidenceType = EvidenceType.other
    title: str = ""
    source: str = ""
    source_type: str
    document_name: str
    status: Literal["available", "partial", "missing", "stale", "not_applicable"]
    content_type: str
    file_name: str
    file_size: int
    storage_path: str | None = None
    preview_text: str
    extracted_signals: list[EvidenceExtractedSignal]
    extraction_status: str = "pending"
    confidence_impact: str = "medium"
    risk_impact: str = "medium"
    related_score_components: list[str]
    lending_question: str = ""
    source_mapping: list[str]
    uploaded_by: str
    reviewed_at: datetime | None = None
    audit_event_id: str | None = None
    created_at: datetime
    updated_at: datetime


class EvidenceUploadResponse(BaseModel):
    record: EvidenceRecord
    audit_event_id: str


class EvidenceStatusUpdate(BaseModel):
    status: Literal["available", "partial", "missing", "stale", "not_applicable"]


class EvidenceMapRow(BaseModel):
    source_type: str
    source_label: str
    source_status: str
    derived_signal: str
    score_component: str
    lending_question: str
    recommended_action: str
    confidence_impact: str
    risk_impact: str


class RecommendedHumanAction(BaseModel):
    label: str
    detail: str
    source: str


class AuditSummary(BaseModel):
    latest_events: list[AuditEventSchema]
    total_events: int


class CreditFileResponse(BaseModel):
    profile: MSMEDetail
    score: ScoreOutputSchema
    prospect: ProspectSignalOutputSchema
    evidence_status: list[EvidenceStatusItem]
    evidence_records: list[EvidenceRecord] = []
    missing_evidence: list[str]
    transaction_summary: TransactionSummary
    risk_warnings: list[EarlyWarningTrigger]
    suggested_credit_posture: str
    recommended_human_actions: list[RecommendedHumanAction]
    audit_summary: AuditSummary
    cited_source_ids: list[str]
    generated_at: datetime


class EvidenceMapResponse(BaseModel):
    msme_id: str
    rows: list[EvidenceMapRow]
    generated_at: datetime


class CaseInboxLane(BaseModel):
    lane: Literal["ready_for_review", "missing_evidence", "risk_attention", "high_potential", "low_confidence"]
    label: str
    cases: list[PortfolioCase]


class CaseInboxResponse(BaseModel):
    lanes: list[CaseInboxLane]
    generated_at: datetime


class CopilotChatRequest(BaseModel):
    message: str
    mode: str | None = None
    include_trace: bool = True


class CopilotChatResponse(BaseModel):
    answer_markdown: str
    decision_support_only: Literal[True] = True
    cited_internal_inputs: list[str]
    trace: list[TraceStep]
    provider: str
    model: str
    prompt_version: str = ""
    summary: str = ""
    recommended_human_action: str = ""
    assumptions: list[str] = []
    follow_up_questions: list[str] = []
    created_at: datetime

File: apps/web/src/lib/schemas/credit-file.ts
import { z } from "zod";
import { auditEventSchema } from "./audit";
import { copilotBriefSchema, traceStepSchema } from "./copilot";
import { msmeDetailSchema } from "./msme";
import { portfolioCaseSchema } from "./portfolio";
import { prospectOutputSchema } from "./prospect";
import { earlyWarningTriggerSchema, scoreOutputSchema } from "./score";

export const transactionSummarySchema = z.object({
  msme_id: z.string(),
  snapshot_month: z.string(),
  monthly_revenue_avg: z.number(),
  monthly_expense_avg: z.number(),
  net_monthly_surplus_estimate: z.number(),
  average_bank_balance: z.number(),
  emi_to_revenue_ratio: z.number(),
  cash_inflow_volatility: z.number(),
  digital_payment_ratio: z.number(),
  bounce_count_3m: z.number(),
  invoice_delay_avg_days: z.number(),
  notes: z.array(z.string())
});

export const evidenceStatusItemSchema = z.object({
  source_type: z.string(),
  source_label: z.string(),
  status: z.string(),
  why_it_matters: z.string(),
  related_score_component: z.string(),
  action_label: z.string(),
  action_enabled: z.boolean(),
  disabled_reason: z.string().nullable()
});

export const evidenceRecordSchema = z.object({
  id: z.string(),
  msme_id: z.string(),
  evidence_type: z.string(),
  title: z.string(),
  source: z.string(),
  source_type: z.string(),
  document_name: z.string(),
  status: z.string(),
  content_type: z.string(),
  file_name: z.string(),
  file_size: z.number(),
  storage_path: z.string().nullable().optional(),
  preview_text: z.string(),
  extracted_signals: z.array(z.object({
    field_name: z.string(),
    value: z.string(),
    source_mapping: z.string(),
    confidence: z.number()
  })),
  extraction_status: z.string(),
  confidence_impact: z.string(),
  risk_impact: z.string(),
  related_score_components: z.array(z.string()),
  lending_question: z.string(),
  source_mapping: z.array(z.string()),
  uploaded_by: z.string(),
  reviewed_at: z.string().nullable().optional(),
  audit_event_id: z.string().nullable().optional(),
  created_at: z.string(),
  updated_at: z.string()
});

export const recommendedHumanActionSchema = z.object({
  label: z.string(),
  detail: z.string(),
  source: z.string()
});

export const creditFileSchema = z.object({
  profile: msmeDetailSchema,
  score: scoreOutputSchema,
  prospect: prospectOutputSchema,
  evidence_status: z.array(evidenceStatusItemSchema),
  evidence_records: z.array(evidenceRecordSchema),
  missing_evidence: z.array(z.string()),
  transaction_summary: transactionSummarySchema,
  risk_warnings: z.array(earlyWarningTriggerSchema),
  suggested_credit_posture: z.string(),
  recommended_human_actions: z.array(recommendedHumanActionSchema),
  audit_summary: z.object({
    latest_events: z.array(auditEventSchema),
    total_events: z.number()
  }),
  cited_source_ids: z.array(z.string()),
  generated_at: z.string()
});

export const evidenceMapRowSchema = z.object({
  source_type: z.string(),
  source_label: z.string(),
  source_status: z.string(),
  derived_signal: z.string(),
  score_component: z.string(),
  lending_question: z.string(),
  recommended_action: z.string(),
  confidence_impact: z.string(),
  risk_impact: z.string()
});

export const evidenceMapResponseSchema = z.object({
  msme_id: z.string(),
  rows: z.array(evidenceMapRowSchema),
  generated_at: z.string()
});

export const caseInboxLaneSchema = z.object({
  lane: z.enum(["ready_for_review", "missing_evidence", "risk_attention", "high_potential", "low_confidence"]),
  label: z.string(),
  cases: z.array(portfolioCaseSchema)
});

export const caseInboxResponseSchema = z.object({
  lanes: z.array(caseInboxLaneSchema),
  generated_at: z.string()
});

export const copilotChatResponseSchema = z.object({
  answer_markdown: z.string(),
  decision_support_only: z.literal(true),
  cited_internal_inputs: z.array(z.string()),
  trace: z.array(traceStepSchema),
  provider: z.string(),
  model: z.string(),
  prompt_version: z.string(),
  summary: z.string(),
  recommended_human_action: z.string(),
  assumptions: z.array(z.string()),
  follow_up_questions: z.array(z.string()),
  created_at: z.string()
});

export type CreditFile = z.infer<typeof creditFileSchema>;
export type EvidenceStatusItem = z.infer<typeof evidenceStatusItemSchema>;
export type EvidenceRecord = z.infer<typeof evidenceRecordSchema>;
export type EvidenceMapRow = z.infer<typeof evidenceMapRowSchema>;
export type EvidenceMapResponse = z.infer<typeof evidenceMapResponseSchema>;
export type CaseInboxResponse = z.infer<typeof caseInboxResponseSchema>;
export type CaseInboxLane = z.infer<typeof caseInboxLaneSchema>;
export type CopilotChatResponse = z.infer<typeof copilotChatResponseSchema>;
export type CopilotBrief = z.infer<typeof copilotBriefSchema>;
